pallets/marketplace/src/lib.rs

```diff
@@ -0,0 +1,486 @@
+			Option<(T::AccountId, u8)>,
+			BalanceOf<T>,
+		),
+		/// Token listed on Marketplace \[owner, class_id, instance_id, author royalty\]
+		TokenListed(T::AccountId, T::NftClassId, T::NftInstanceId, T::AccountId, u8),
+		/// Token listed on Marketplace \[class_id, instance_id\]
+		TokenUnlisted(T::NftClassId, T::NftInstanceId),
+		/// Offer was placed on a token \[offerer, class_id, instance_id, price\]
+		OfferPlaced(T::AccountId, T::NftClassId, T::NftInstanceId, BalanceOf<T>),
+		/// Offer was withdrawn \[sender, class_id, instance_id\]
+		OfferWithdrawn(T::AccountId, T::NftClassId, T::NftInstanceId),
+		/// Offer was accepted \[sender, class_id, instance_id\]
+		OfferAccepted(T::AccountId, T::NftClassId, T::NftInstanceId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Account is not the owner of the token
+		NotTheTokenOwner,
+		/// Cannot buy a token from yourself
+		BuyFromSelf,
+		/// Token is currently not for sale
+		NotForSale,
+		/// This class type cannot be listed on Marketplace
+		UnsupportedClassType,
+		/// Token is not listed on Marketplace
+		NotListed,
+		/// Class or instance does not exist
+		ClassOrInstanceUnknown,
+		/// Token already listed on marketplace
+		AlreadyListed,
+		/// Offer is lower than the pallet limit
+		OfferTooLow,
+		/// No offer for this token found from the user
+		UnknownOffer,
+		/// Offer is no longer valid
+		OfferExpired,
+		/// User already offered for this token
+		AlreadyOffered,
+		/// User has to be offer maker or token owner to withdraw an offer
+		WithdrawNotAuthorized,
+		/// User has to be the token owner to accept an offer
+		AcceptNotAuthorized,
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	// Call extrinsic helper function used by `buy` and `accept_offer` functions
+	fn do_buy(
+		buyer: T::AccountId,
+		class_id: T::NftClassId,
+		instance_id: T::NftInstanceId,
+		is_offer: bool,
+	) -> DispatchResult {
+		let owner = pallet_uniques::Pallet::<T>::owner(class_id.into(), instance_id.into())
+			.ok_or(Error::<T>::ClassOrInstanceUnknown)?;
+		ensure!(buyer != owner, Error::<T>::BuyFromSelf);
+
+		let owner_origin = T::Origin::from(RawOrigin::Signed(owner.clone()));
+
+		pallet_uniques::Pallet::<T>::thaw(owner_origin, class_id.into(), instance_id.into())?;
+
+		let token_id = (class_id, instance_id);
+
+		Tokens::<T>::try_mutate(class_id, instance_id, |maybe_token_info| -> DispatchResult {
+			let token_info = maybe_token_info.as_mut().ok_or(Error::<T>::NotListed)?;
+
+			let mut price;
+
+			if is_offer {
+				price = Offers::<T>::get(token_id, buyer.clone())
+					.map(|o| o.amount)
+					.ok_or(Error::<T>::UnknownOffer)?;
+			} else {
+				price = token_info.price.take().ok_or(Error::<T>::NotForSale)?;
+			}
+
+			let royalty = Instances::<T>::get(class_id, instance_id)
+				.map(|i| i.royalty)
+				.ok_or(pallet_nft::Error::<T>::RoyaltyNotSet)?;
+			let author = Instances::<T>::get(class_id, instance_id)
+				.map(|i| i.author)
+				.ok_or(pallet_nft::Error::<T>::AuthorNotSet)?;
```

Comment:
Same here as in previous case - get the data first and just ensure royalty and author.